README.md
# ROS2 Ouster Drivers

These are an implementation of ROS2 drivers for the Ouster lidar. This includes all models of the OS-x from 16 to 128 beams running the firmware 2.1.

You can find a few videos looking over the sensor below. They both introduce the ROS1 driver but are extremely useful references regardless:

OS-1 Networking Setup      |  OS-1 Data Overview
:-------------------------:|:-------------------------:
[![IMAGE ALT TEXT HERE](http://img.youtube.com/vi/92ajXjIxDGM/0.jpg)](http://www.youtube.com/watch?v=92ajXjIxDGM) | [![IMAGE ALT TEXT HERE](http://img.youtube.com/vi/4VgGG8Xe4IA/0.jpg)](http://www.youtube.com/watch?v=4VgGG8Xe4IA)


## Documentation

Documentation can be generated using Doxygen.

Run `doxygen` in the root of this repository. It will generate a `/doc/*` directory containing the documentation. Entrypoint in a browser is `index.html`.

## Design

See design doc in `design/*` directory [here](ros2_ouster/design/design_doc.md).

## ROS Interfaces

<center>

| Topic                | Type                    | Description                                      |
|----------------------|-------------------------|--------------------------------------------------|
| `scan`               | sensor_msgs/LaserScan   | 2D laser scan of the 0-angle ring                |
| `range_image`        | sensor_msgs/Image       | Image of the range values from the sensor        |
| `intensity_image`    | sensor_msgs/Image       | Image of the Intensity values from the sensor    |
| `noise_image`        | sensor_msgs/Image       | Image of the noise values from the sensor        |
| `reflectivity_image` | sensor_msgs/Image       | Image of the reflectivity values from the sensor |
| `points`             | sensor_msgs/PointCloud2 | 3D Pointcloud generated from a 360 rotation      |
| `imu`                | sensor_msgs/Imu         | IMU values at transmission rate                  |

| Service           | Type                    | Description                       |
|-------------------|-------------------------|-----------------------------------|
| `reset`           | std_srvs/Empty          | Reset the sensor's connection     |
| `GetMetadata`     | ouster_msgs/GetMetadata | Get information about the sensor. A optional filepath can be specified to save the metadata to a local file.  |

| Parameter                | Type    | Description                                                                                                 |
|--------------------------|---------|-------------------------------------------------------------------------------------------------------------|
| `lidar_ip`               | String  | IP or hostname of lidar (ex. 10.5.5.87, os1-serialno.local)                                                 |
| `computer_ip`            | String  | IP or hostname of computer to get data (ex. 10.5.5.1) or broadcast (ex. 255.255.255.255) or if using the default driver, "" for automatic detection                   |
| `lidar_mode`             | String  | Mode of data capture, default `512x10`                                                                      |
| `imu_port`               | int     | Port of IMU data, default 7503                                                                              |
| `lidar_port`             | int     | Port of laser data, default 7502                                                                            |
| `sensor_frame`           | String  | TF frame of sensor, default `laser_sensor_frame`                                                            |
| `laser_frame`            | String  | TF frame of laser data, default `laser_data_frame`                                                          |
| `imu_frame`              | String  | TF frame of imu data, default `imu_data_frame`                                                              |
| `ethernet_device`        | String  | An ethernet device (e.g. eth0 or eno1) on which the Tins driver will listen for packets. Note that this is only a parameter for the Tins driver, and is only specified in the config file for that driver. |                              |
| `use_system_default_qos` | bool    | Publish data with default QoS for rosbag2 recording, default `False`                                        |
| `timestamp_mode`         | String  | Method used to timestamp measurements, default `TIME_FROM_INTERNAL_OSC`                                     |
| `os1_proc_mask`          | String  | Mask encoding data processors to activate, default <code>IMG &#124; PCL &#124; IMU &#124; SCAN</code> |
| `pointcloud_filter_zero_points` | bool | Reduce pointcloud size by omitting (0, 0, 0) points, default `False`. If used, will make the PC2 unstructured.     |


</center>

Note: TF will provide you the transformations from the sensor frame to each of
the data frames.

### Timestamp Modes

Referring to the parameter table above, the `timestamp_mode` parameter has four
allowable options (as of this writing). They are: `TIME_FROM_INTERNAL_OSC`,
`TIME_FROM_SYNC_PULSE_IN`, `TIME_FROM_PTP_1588`, `TIME_FROM_ROS_RECEPTION`. A
description of each now follows.

#### `TIME_FROM_INTERNAL_OSC`

Use the LiDAR internal clock. Measurements are time stamped with ns since
power-on. Free running counter based on the sensor’s internal oscillator. Counts
seconds and nanoseconds since sensor turn on, reported at ns resolution (both a
second and nanosecond register in every UDP packet), but min increment is on
the order of 10 ns. Accuracy is +/- 90 ppm.

#### `TIME_FROM_SYNC_PULSE_IN`

A free running counter synced to the `SYNC_PULSE_IN` input counts seconds (# of
pulses) and nanoseconds since sensor turn on. If `multipurpose_io_mode` is set to
`INPUT_NMEA_UART` then the seconds register jumps to time extracted from a NMEA
`$GPRMC` message read on the `multipurpose_io` port. Reported at ns resolution
(both a second and nanosecond register in every UDP packet), but min increment
is on the order of 10 ns. Accuracy is +/- 1 s from a perfect `SYNC_PULSE_IN`
source.

#### `TIME_FROM_PTP_1588`

Synchronize with an external PTP master. A monotonically increasing counter
that will begin counting seconds and nanoseconds since startup. As soon as a
1588 sync event happens, the time will be updated to seconds and nanoseconds
since 1970. The counter must always count forward in time. If another 1588 sync
event happens the counter will either jump forward to match the new time, or
slow itself down. It is reported at ns resolution (there is both a second and
nanosecond register in every UDP packet), but the minimum increment
varies. Accuracy is +/- <50 us from the 1588 master.

#### `TIME_FROM_ROS_RECEPTION`

The sensor will run in `TIME_FROM_INTERNAL_OSC` time mode but data are stamped with the ROS time when they are received. The inherent latency
between when the data were sampled by the LiDAR and when the data were received
by this ROS node is not modelled. This approach may be acceptable to get up and
running quickly or for static applications. However, for mobile robots,
particularly those traveling at higher speeds, it is not recommended to use
this `timestamp_mode`. When running in this mode, the on-LiDAR `timestamp_mode`
will not be set by this driver.

### Parameterizing the Active Data Processors

The `os1_proc_mask` parameter is set to a mask-like-string used to define the
data processors that should be activated upon startup of the driver. This will
determine the topics that are available for client applications to consume. The
*de facto* reference for these values are defined in
[processor_factories.hpp](ros2_ouster/include/ros2_ouster/processors/processor_factories.hpp). It
is recommended to only use the processors that you require for your application.

The available data processors are:

- **IMG** Provides 8-bit image topics encoding the noise, range, intensity, and
  reflectivitiy from a scan.
- **PCL** Provides a point cloud encoding of a LiDAR scan
- **IMU** Provides a data stream from the LiDAR's integral IMU
- **SCAN** Provides a 2D LaserScan from the closest to 0-degree azimuth ring

To construct a valid string for the `os1_proc_mask` parameter, join the tokens
from above (in any combination) with the pipe character (`|`). For example,
valid strings include but are not limited to: `IMG|PCL`, `IMG|PCL|IMU`, `PCL`,
etc. The default value is `IMG|PCL|IMU|SCAN`.

More details about data processors in the driver is provided in the [Additional
Lidar Processing](#additional-lidar-processing) section below.

## Extensions

This package was intentionally designed for new capabilities to be added. Whether that being supporting new classes of Ouster lidars (sensor-custom, OS2, ...) or supporting new ways of processing the data packets.

### Additional Lidar Processing
It can be imagined that if you have a stream of lidar or IMU packets, you may want to process them differently. If you're working with a high speed vehicle, you may want the packets projected into a pointcloud and published with little batching inside the driver. If you're working with pointclouds for machine learning, you may only want the pointcloud to include the `XYZ` information and not the intensity, reflectivity, and noise information to reduce dimensionality.

In any case, I provide a set of logical default processing implementations on the lidar and IMU packets. These are implementations of the `ros2_ouster::DataProcessorInterface` class in the `interfaces` directory. To create your own processor to change the pointcloud type, buffering methodology, or some new cool thing, you must create an implementation of a data processor.

After creating your implementation, that will take in a `uint8_t *` of a data packet and accomplish your task, you will need to create a factory method for it in `processor_factories.hpp` and add it to the list of processors to be created in the `createProcessors` method.

I encourage you to contribute back any new processor methods to this project! The default processors will buffer 1 full rotation of data of the pointcloud and publish the pointcloud with the X, Y, Z, range, intensity, reflectivity, ring, and noise information. It will also buffer a full rotation and publish the noise, intensity, and reflectivity images. Finally, it will publish the IMU data at transmission frequency.

Some examples:
- If you wanted the points at transmission frequency to reduce aliasing
- Different types of pointclouds published containing a subset or additional information.
- If you wanted the information in another format (ei 1 data image with 3 channels of the range, intensity, and noise)
- Downsample the data at a driver level to only take every `N`th ring.

### Additional Lidar Units
To create a new lidar for this driver, you only need to make an implementation of the `ros2_ouster::SensorInterface` class and include any required SDKs. Then, in the `driver_types.hpp` file, add your new interface as a template of the `OusterDriver` and you're good to go.

You may need to add an additional `main` method for the new templated program, depending if you're using components. If it uses another underlying SDK other than `sensor` you will also need to create new processors for it as the processors are bound to a specific unit as the data formatting may be different. If they are the same, you can reuse the `sensor` processors.

## Lifecycle

This ROS2 driver makes use of Lifecycle nodes. If you're not familiar with lifecycle, or managed nodes, please familiarize yourself with the [ROS2 design](https://design.ros2.org/articles/node_lifecycle.html) document on it.

The lifecycle node allow for the driver to have a deterministic setup and tear down and is a new feature in ROS2. The launch script will need to use a lifecycle node launcher to transition into the active state to start processing data.

## Component

This ROS2 driver makes use of Component nodes. If you're not familiar with component nodes please familiarize yourself with the [ROS2 design](https://index.ros.org/doc/ros2/Tutorials/Composition) document on it.

The component node allow for the driver and its processing nodes to be launched into the same process and is a new feature in ROS2. This allows the sensor and its data clients to operate without serialization or copying between nodes sharing a memory pool.

There's a little work in ROS2 Eloquent to launch a component-lifecycle node using only the roslaunch API. It may be necessary to include the Ouster driver in your lifecycle manager to transition into the active state when loading the driver into a process as a component. Example using the ROS2 component manager:

```
# component manager for dynamic loading (also may be done through launch)
ros2 run rclcpp_components component_container
# load this component
ros2 component load /ComponentManager ros2_ouster ros2_ouster::OS1Driver
# Set parameters
ros2 param set OusterDriver lidar_ip 10.5.5.86
ros2 param set OusterDriver computer_ip 10.5.5.1
# transition to configuring lifecycle stage
ros2 lifecycle set OusterDriver 1
# transition to active lifecycle stage (will now stream data)
ros2 lifecycle set OusterDriver 3
```

# Ouster Messages

A message `Metadata` was created to describe the metadata of the lidars. In addition the `GetMetadata` service type will be used to get the metadata from a running driver.


# Setup and Networking

Ouster gives you some tools to set up a direct connection to the sensor from you computer. I'd argue these are a bit obtuse and they should really provide a set of scripts to set this up automatically as a daemon. However, since I am also using this as a development tool, I don't want this always running in the background on my machines so I provide the directions below to setup the network connection.

### One time setup with IPv4

These are bash commands in Linux to setup the connection. These steps only need to happen the first time you set up the laser. After the first time, when you establish the network connection to the sensor, you can just select this created network profile. **Ensure the sensor is powered off and disconnected at this point.**

The `[eth name]` is the nework interface you're connecting to. On older Linux systems, that's `eth0` or similar. In newer versions, its `enp...` or `enx...` when you look at the output of `ifconfig`.

```
 ip addr flush dev [eth name]
 ip addr show dev [eth name]
```

The output you see from `show` should look something like `[eth name] ... state DOWN ...`. Its only important that you see `DOWN` and not `UP`. Next, lets setup a static IP address for your machine so you can rely on this in the future. Ouster uses the 10.5.5.* range, and I don't see a compelling reason to argue with it.

```
sudo ip addr add 10.5.5.1/24 dev [eth name]

```

Now, lets setup the connection. At this point you may now plug in and power on your sensor.

```
sudo ip link set [eth name] up
sudo addr show dev [eth name]
```

The output you see from `show` should look something like `[eth name] ... state UP ...`. Its only important that you see `UP` now and not `DOWN`. At this point, you've setup the networking needed for the one time setup.

### Connection with IPv4

We can setup the network connection to the sensor now with the proper settings. Note: This command could take up to 30 seconds to setup, be patient. If after a minute you see no results, then you probably have an issue. Start the instructions above over. Lets set up the network

```
sudo dnsmasq -C /dev/null -kd -F 10.5.5.50,10.5.5.100 -i [eth name] --bind-dynamic
```

Instantly you should see something similar to:

```
dnsmasq: started, version 2.75 cachesize 150
dnsmasq: compile time options: IPv6 GNU-getopt DBus i18n IDN DHCP DHCPv6 no-Lua TFTP conntrack ipset auth DNSSEC loop-detect inotify
dnsmasq-dhcp: DHCP, IP range 10.5.5.50 -- 10.5.5.100, lease time 1h
dnsmasq-dhcp: DHCP, sockets bound exclusively to interface enxa0cec8c012f8
dnsmasq: reading /etc/resolv.conf
dnsmasq: using nameserver 127.0.1.1#53
dnsmasq: read /etc/hosts - 10 addresses

```

You need to wait until you see something like:

```
dnsmasq-dhcp: DHCPDISCOVER(enxa0cec8c012f8) [HWaddr]
dnsmasq-dhcp: DHCPOFFER(enxa0cec8c012f8) 10.5.5.87 [HWaddr]
dnsmasq-dhcp: DHCPREQUEST(enxa0cec8c012f8) 10.5.5.87 [HWaddr]
dnsmasq-dhcp: DHCPACK(enxa0cec8c012f8) 10.5.5.87 [HWaddr] os1-SerialNumXX
```

Now you're ready for business. Lets see what IP address it's on (10.5.5.87). Lets ping it

```
ping 10.5.5.87
```

and we're good to go!

### Using IPv6 with link local

Instead of having to configure `dnsmasq` and static addresses in the previous section, you can use link local IPv6 addresses.

1. Find the link local address of the Ouster. With avahi-browse, we can find the address of the ouster by browsing all non-local services and resolving them.
```
$ avahi-browse -arlt
+   eth2 IPv6 Ouster Sensor 992109000xxx                    _roger._tcp          local
+   eth2 IPv4 Ouster Sensor 992109000xxx                    _roger._tcp          local
=   eth2 IPv6 Ouster Sensor 992109000xxx                    _roger._tcp          local
   hostname = [os-992109000xxx.local]
   address = [fe80::be0f:a7ff:fe00:2861]
   port = [7501]
   txt = ["fw=ousteros-image-prod-aries-v2.0.0+20201124065024" "sn=992109000xxx" "pn=840-102144-D"]
=   eth2 IPv4 Ouster Sensor 992109000xxx                    _roger._tcp          local
   hostname = [os-992109000xxx.local]
   address = [192.168.90.2]
   port = [7501]
   txt = ["fw=ousteros-image-prod-aries-v2.0.0+20201124065024" "sn=992109000xxx" "pn=840-102144-D"]

```
As shown above, on interface `eth2`, the ouster has an IPv6 link local address of `fe80::be0f:a7ff:fe00:2861`.

To use link local addressing with IPv6, the standard way to add a scope ID is appended with a `%` character like so in `sensor.yaml`. Automatic detection for computer IP address can be used with an empty string.
```
lidar_ip: "fe80::be0f:a7ff:fe00:2861%eth2"
computer_ip: ""
```

Note that this feature is only available with the default driver version, configured by `driver_config.yaml`. When running the Tins-based driver (see the following sections), both the LiDAR and computer IP address must be specified in `tins_driver_config.yaml`.

### Usage with the default driver

Now that we have a connection over the network, lets view some data. After building your colcon workspace with this package, source the install space, then run:

```
ros2 launch ros2_ouster driver_launch.py
```

Make sure to update your parameters file if you don't use the default IPs (10.5.5.1, 10.5.5.87). You may also use the `.local` version of your ouster lidar. To find your IPs, see the `dnsmasq` output or check with `nmap -SP 10.5.5.*/24`.
An alternative tool is [avahi-browse](https://linux.die.net/man/1/avahi-browse): 

```
avahi-browse -arlt
```

Now that your connection is up, you can view this information in RViz. Open an RViz session and subscribe to the points, images, and IMU topics in the laser frame. When trying to visualize the point clouds, be sure to change the Fixed Frame under Global Options to "laser_data_frame" as this is the default parent frame of the point cloud headers.

When the driver configures itself, it will automatically read the metadata parameters from the Ouster. If you wish to save these parameters and use them with captured data (see the next section) then you can save the data to a specific location using the `getMetadata` service that the driver offers. To use it, run the driver with a real Ouster LiDAR and then make the following service call:

```
ros2 service call /ouster_driver/get_metadata ouster_msgs/srv/GetMetadata "{metadata_filepath: "/path/to/your/metadata.json"}"
```

The driver will then save all the required metadata to the specified file and dump the same metadata as a JSON string to the terminal. Alternatively, the service can be called without specifying a filepath (see below) in which case no file will be saved, and the metadata will still be printed to terminal. Copying this string and manually saving it to a .json file is also a valid way to generate a metadata file.  

```
ros2 service call /ouster_driver/get_metadata ouster_msgs/srv/GetMetadata
```

Have fun!

### Usage with Tins-based driver

If you want to use the driver to read data from a pcap file, you can use the `Tins`-based driver. To do this, open the `tins_driver_config.yaml` file and edit the following parameter:

* `ethernet_device`: Change this to a working ethernet device on your computer that you plan to replay data through (e.g. "eth1").

You can run the Tins driver with the command below. This will use the default `ouster_os0128_1024_metadata.json` file:

```
ros2 launch ros2_ouster tins_driver_launch.py
```

Alternatively, you can change the metadata being used by specifying the metadata filepath as shown below. You can generate a metadata file using the `getMetadata` service as shown in the previous section. Or you can use one of the example metadata files provided, which come from an Ouster OS0-128 in either 1024x10 or 2048x10 mode.

```
ros2 launch ros2_ouster tins_driver_launch.py metadata_filepath:=/path/to/metadata.json
```

After launching the driver, in a new terminal, you can replay a pcap file of recorded ouster data using the following command (as an example):

```
sudo tcpreplay --intf1=eth1 saved_ouster_data.pcap 
```

You may need to run this command with `sudo`. Note that this driver version will also work with a live Ouster sensor, provided the data is coming into the correct ethernet device, and the parameters in the metadata file match those of the sensor. However it is recommended that you run the default drive with a real sensor, as this will guarantee that the metadata settings are correct.

CHANGELOG.rst
Changelog
=========

[0.3.0] Added a new SensorTins implementation (2021-06-23)
-------------

**General Changes**

* Added example metadata files for OS0-128 in 1024x10 and 2048x10 mode.
* Added/exposed the functions for loading and saving metadata to a json file.
* Added metadatat file saving to ``getMetadata`` service.
* Contibutors: Matthew Young (Trimble Inc).

**TinsDriver**

* Added a new SensorImplementation called SensorTins, and a Driver implementation that uses it called TinsDriver. This implementation uses the Tins library to find LiDAR and IMU packets, and is intended to be used with replayed pcap files rather than a real LiDAR. 
* Added a new driver called TinsDriver which utilizes the existing OusterDriver class with the SensorTins class.
* Added an new set of launch, param and main files for the TinsDriver.
* Updated readme and parameter file to explain usage of the TinsDriver implementation.
* Added some additional debug code to print the number of packets that have been accumulated into a cloud when it is published.
* Contibutors: Matthew Young (Trimble Inc).

[0.3.0] Ported client changes for FW 2.0/2.1 (2021-06-23)
-------------

**ouster_client**

* From ouster_example (`#239 <https://github.com/ouster-lidar/ouster_example/commit/e1176f427f68eb0807bb15ccabe34aea47a1c5d3>`_). 
    * Optional header and license file are now present in their own folder: ``include/ros2_ouster/client/optional-lite/``.
    * Ported some mode-getting functions that use Optional to types.cpp.
    * Ported sensor_config struct to types.hpp.
    * Ported the get/set_config functions for sensor_config to client.cpp/.hpp.
    * Ported some documentation updates, and a != operator for LidarScan objects to lidar_scan.h.
    * Ported several additional enumerated types (e.g. OperatingMode and MultipurposeIOMode) to types.h.
    * Ported parsing functions for additional enumerated types to types.cpp.
    * Ported ColumnWindow definition and added it as a parameter to data_format struct, which is the only existing struct changed.
    * In lidar_scan.cpp, ported some changes to ``make_xyz_lut`` and ``cartesian`` functions.  
* From ouster_example (`#246 <https://github.com/ouster-lidar/ouster_example/commit/2b49e6a2f3dbd0462c557974a3f428915067fd2f>`_).
    * Removed a json include incorrectly added to types.h.  
* From ouster_example (`#259 <https://github.com/ouster-lidar/ouster_example/commit/b8b23c35d7b719d69341a438d386a801688aa6a4#diff-db793e44a91d87bc6e0d94870833d5d9eeb4420e2fbbec4e2468717a359a651f>`_).
    * Increased default timeout time from 30 to 60 seconds for ``init_client`` and ``read_imu_packet``.
    * In netcompat.cpp, changed non_blocking_mode from off to on (0 to 1) if using Win32.
    * Added ``signal_multiplier`` parameter to ``sensor_config`` in types.h.
    * Added ``signal_multiplier`` parameter setting to some functions in client.cpp and types.cpp.
* Updated changelog.

**General Changes**

* Changed indenting in client/types/lidar_scan source header files to be more consistent with existing indenting style.

[0.3.0] (YYYY-MM-DD)
------------------
* Increased the rate at which the timer callback fires to process incoming data
  from 1280 Hz to 2560 Hz.
  (`#55 <https://github.com/SteveMacenski/ros2_ouster_drivers/issues/55>`_)
* Reordered in-memory data of point cloud data processor to row-major order
  (`#52 <https://github.com/SteveMacenski/ros2_ouster_drivers/issues/52>`_)
* Start tracking changes in CHANGELOG at 0.3.0
* Contributors: Steve Macenski, Tom Panzarella


ros2_ouster/src/ouster_driver.cpp
// Copyright 2021, Steve Macenski
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <chrono>
#include <memory>
#include <string>
#include <utility>
#include <vector>

#include "rclcpp/qos.hpp"
#include "ros2_ouster/exception.hpp"
#include "ros2_ouster/interfaces/lifecycle_interface.hpp"
#include "ros2_ouster/interfaces/sensor_interface.hpp"
#include "ros2_ouster/ouster_driver.hpp"
#include "ros2_ouster/processors/processor_factories.hpp"
#include "ros2_ouster/client/types.h"
#include "ros2_ouster/sensor.hpp"
#include "ros2_ouster/sensor_tins.hpp"

namespace ros2_ouster
{

using std::placeholders::_1;
using std::placeholders::_2;
using std::placeholders::_3;
using namespace std::chrono_literals;

OusterDriver::OusterDriver(
  std::unique_ptr<SensorInterface> sensor,
  const rclcpp::NodeOptions & options)
: LifecycleInterface("OusterDriver", options), _sensor{std::move(sensor)}
{
  // Declare parameters for configuring the _driver_
  this->declare_parameter("sensor_frame", std::string("laser_sensor_frame"));
  this->declare_parameter("laser_frame", std::string("laser_data_frame"));
  this->declare_parameter("imu_frame", std::string("imu_data_frame"));
  this->declare_parameter("use_system_default_qos", false);
  this->declare_parameter("proc_mask", std::string("IMG|PCL|IMU|SCAN"));

  // Declare parameters used across ALL _sensor_ implementations
  this->declare_parameter<std::string>("lidar_ip");
  this->declare_parameter<std::string>("computer_ip");
  this->declare_parameter("imu_port", 7503);
  this->declare_parameter("lidar_port", 7502);
  this->declare_parameter("lidar_mode", std::string("512x10"));
  this->declare_parameter("timestamp_mode", std::string("TIME_FROM_INTERNAL_OSC"));
}

OusterDriver::~OusterDriver() = default;

void OusterDriver::onConfigure()
{
  // Get parameters for configuring the _driver_
  _laser_sensor_frame = get_parameter("sensor_frame").as_string();
  _laser_data_frame = get_parameter("laser_frame").as_string();
  _imu_data_frame = get_parameter("imu_frame").as_string();
  _use_system_default_qos = get_parameter("use_system_default_qos").as_bool();
  _proc_mask = ros2_ouster::toProcMask(get_parameter("proc_mask").as_string());

  // Get parameters used across ALL _sensor_ implementations. Parameters unique
  // a specific Sensor implementation are "getted" in the configure() function
  // for that sensor.
  ros2_ouster::Configuration lidar_config;
  lidar_config.imu_port = this->get_parameter("imu_port").as_int();
  lidar_config.lidar_port = this->get_parameter("lidar_port").as_int();
  lidar_config.lidar_mode = this->get_parameter("lidar_mode").as_string();
  lidar_config.timestamp_mode = this->get_parameter("timestamp_mode").as_string();

  // Deliberately retrieve the IP parameters in a try block without defaults, as
  // we cannot estimate a reasonable default IP address for the LiDAR/computer.
  try {
    lidar_config.lidar_ip = get_parameter("lidar_ip").as_string();
    lidar_config.computer_ip = get_parameter("computer_ip").as_string();
  } catch (...) {
    RCLCPP_FATAL(
      this->get_logger(),
      "Failed to get lidar or IMU IP address or "
      "hostname. An IP address for both are required!");
    exit(-1);
  }

  if (lidar_config.timestamp_mode == "TIME_FROM_ROS_RECEPTION") {
    RCLCPP_WARN(
      this->get_logger(),
      "Using TIME_FROM_ROS_RECEPTION to stamp data with ROS time on "
      "reception. This has unmodelled latency!");
    lidar_config.timestamp_mode = "TIME_FROM_INTERNAL_OSC";
    _use_ros_time = true;
  } else {
    _use_ros_time = false;
  }

  // Configure the driver and sensor
  try {
    _sensor->configure(lidar_config, shared_from_this());
  } catch (const OusterDriverException & e) {
    RCLCPP_FATAL(this->get_logger(), "Exception thrown: (%s)", e.what());
    exit(-1);
  }

  RCLCPP_INFO(
    this->get_logger(),
    "This driver is compatible with sensors running fw 2.x.");

  _reset_srv = this->create_service<std_srvs::srv::Empty>(
    "~/reset", std::bind(&OusterDriver::resetService, this, _1, _2, _3));
  _metadata_srv = this->create_service<ouster_msgs::srv::GetMetadata>(
    "~/get_metadata", std::bind(&OusterDriver::getMetadata, this, _1, _2, _3));

  _full_rotation_accumulator = std::make_shared<sensor::FullRotationAccumulator>(
    _sensor->getMetadata(), _sensor->getPacketFormat());

  if (_use_system_default_qos) {
    RCLCPP_INFO(
      this->get_logger(), "Using system defaults QoS for sensor data");
    _data_processors = ros2_ouster::createProcessors(
      shared_from_this(), _sensor->getMetadata(), _imu_data_frame, _laser_data_frame,
      rclcpp::SystemDefaultsQoS(),
      _sensor->getPacketFormat(), _full_rotation_accumulator, _proc_mask);
  } else {
    _data_processors = ros2_ouster::createProcessors(
      shared_from_this(), _sensor->getMetadata(), _imu_data_frame, _laser_data_frame,
      rclcpp::SensorDataQoS(), _sensor->getPacketFormat(), _full_rotation_accumulator, _proc_mask);
  }

  _tf_b = std::make_unique<tf2_ros::StaticTransformBroadcaster>(
    shared_from_this());
  broadcastStaticTransforms(_sensor->getMetadata());
}

void OusterDriver::onActivate()
{
  DataProcessorMapIt it;
  for (it = _data_processors.begin(); it != _data_processors.end(); ++it) {
    it->second->onActivate();
  }

  // Speed of the lidar is 1280 hz. We fire our timer event at 2x that rate to
  // ensure we can process all of the incoming data in a timely manner.
  // See: https://github.com/SteveMacenski/ros2_ouster_drivers/issues/55
  _process_timer = this->create_wall_timer(
    390625ns, std::bind(&OusterDriver::processData, this));
}

void OusterDriver::onError()
{
}

void OusterDriver::onDeactivate()
{
  _process_timer->cancel();
  _process_timer.reset();

  DataProcessorMapIt it;
  for (it = _data_processors.begin(); it != _data_processors.end(); ++it) {
    it->second->onDeactivate();
  }
}

void OusterDriver::onCleanup()
{
  _data_processors.clear();
  _tf_b.reset();
  _reset_srv.reset();
  _metadata_srv.reset();
}

void OusterDriver::onShutdown()
{
  _process_timer->cancel();
  _process_timer.reset();
  _tf_b.reset();

  DataProcessorMapIt it;
  _data_processors.clear();
}

void OusterDriver::broadcastStaticTransforms(
  const ouster::sensor::sensor_info & mdata)
{
  if (_tf_b) {
    std::vector<geometry_msgs::msg::TransformStamped> transforms;
    transforms.push_back(
      toMsg(
        mdata.imu_to_sensor_transform,
        _laser_sensor_frame, _imu_data_frame, this->now()));
    transforms.push_back(
      toMsg(
        mdata.lidar_to_sensor_transform,
        _laser_sensor_frame, _laser_data_frame, this->now()));
    _tf_b->sendTransform(transforms);
  }
}

void OusterDriver::processData()
{
  try {
    ouster::sensor::client_state state = _sensor->get();
    _lidar_packet_data = _sensor->readLidarPacket(state);
    _imu_packet_data = _sensor->readImuPacket(state);

    RCLCPP_DEBUG(
      this->get_logger(),
      "Retrieved packet with state: %s",
      ros2_ouster::toString(state).c_str());

    std::pair<DataProcessorMapIt, DataProcessorMapIt> key_its;

    uint64_t override_ts =
      this->_use_ros_time ? this->now().nanoseconds() : 0;

    if (_lidar_packet_data) {
      _full_rotation_accumulator->accumulate(_lidar_packet_data, override_ts);

      key_its = _data_processors.equal_range(ouster::sensor::client_state::LIDAR_DATA);

      for (DataProcessorMapIt it = key_its.first; it != key_its.second; it++) {
        it->second->process(_lidar_packet_data, override_ts);
      }
    }

    if (_imu_packet_data) {
      key_its = _data_processors.equal_range(ouster::sensor::client_state::IMU_DATA);

      for (DataProcessorMapIt it = key_its.first; it != key_its.second; it++) {
        it->second->process(_imu_packet_data, override_ts);
      }
    }
  } catch (const OusterDriverException & e) {
    RCLCPP_WARN(
      this->get_logger(),
      "Failed to process packet with exception %s.", e.what());
  }
}

void OusterDriver::resetService(
  const std::shared_ptr<rmw_request_id_t>/*request_header*/,
  const std::shared_ptr<std_srvs::srv::Empty::Request> request,
  std::shared_ptr<std_srvs::srv::Empty::Response> response)
{
  if (!this->isActive()) {
    return;
  }

  ros2_ouster::Configuration lidar_config;
  lidar_config.lidar_ip = get_parameter("lidar_ip").as_string();
  lidar_config.computer_ip = get_parameter("computer_ip").as_string();
  lidar_config.imu_port = get_parameter("imu_port").as_int();
  lidar_config.lidar_port = get_parameter("lidar_port").as_int();
  lidar_config.lidar_mode = get_parameter("lidar_mode").as_string();
  lidar_config.timestamp_mode = get_parameter("timestamp_mode").as_string();
  _sensor->reset(lidar_config, shared_from_this());
}

void OusterDriver::getMetadata(
  const std::shared_ptr<rmw_request_id_t>/*request_header*/,
  const std::shared_ptr<ouster_msgs::srv::GetMetadata::Request> request,
  std::shared_ptr<ouster_msgs::srv::GetMetadata::Response> response)
{
  if (!this->isActive()) {
    return;
  }
  response->metadata = toMsg(_sensor->getMetadata());

  // Save the metadata to file ONLY if the user specifies a filepath
  if (request->metadata_filepath != "") {
    std::string json_config = ouster::sensor::to_string(_sensor->getMetadata());
    std::ofstream ofs;
    ofs.open(request->metadata_filepath);
    ofs << json_config << std::endl;
    ofs.close();
    if (!ofs) {
      RCLCPP_ERROR(
        this->get_logger(),
        "Failed to save metadata to: %s.",
        request->metadata_filepath.c_str());
    } else {
      RCLCPP_INFO(
        this->get_logger(),
        "Saving metadata to a .json file specifed here: %s",
        request->metadata_filepath.c_str());
    }
  }
}

}  // namespace ros2_ouster
